server/game/gameSystems/DeckSearchSystem.ts

```diff
@@ -0,0 +1,284 @@
+        const properties = this.generatePropertiesFromContext(context);
+        const amount = this.getAmount(properties.cardsToSearch, context);
+        const message =
+            amount > 0
+                ? `look at the top ${amount === 1 ? 'card' : `${amount} cards`} of their deck`
+                : 'search their deck';
+        return [message, []];
+    }
+
+    public override canAffect(player: Player, context: AbilityContext, additionalProperties = {}): boolean {
+        const properties = this.generatePropertiesFromContext(context, additionalProperties) as IDeckSearchProperties;
+        const amount = this.getAmount(properties.cardsToSearch, context);
+        return amount !== 0 && this.getDeck(player).length > 0 && super.canAffect(player, context);
+    }
+
+    public override defaultTargets(context: AbilityContext): Player[] {
+        return [context.player];
+    }
+
+    public override addPropertiesToEvent(event: DeckSearchEvent, player: Player, context: AbilityContext, additionalProperties: unknown): void {
+        const { cardsToSearch: amount } = this.generatePropertiesFromContext(context, additionalProperties) as IDeckSearchProperties;
+        const fAmount = this.getAmount(amount, context);
+        super.addPropertiesToEvent(event, player, context, additionalProperties);
+        event.amount = fAmount;
+    }
+
+    public override generateEventsForAllTargets(context: AbilityContext, additionalProperties = {}): GameEvent[] {
+        const events: GameEvent[] = [];
+
+        const properties = this.generatePropertiesFromContext(context, additionalProperties) as IDeckSearchProperties;
+        const player = properties.player || context.player;
+        const event = this.generateEvent(player, context, additionalProperties) as DeckSearchEvent;
+        const amount = event.amount > -1 ? event.amount : this.getDeck(player).length;
+        let cards = this.getDeck(player).slice(0, amount);
+        if (event.amount === -1) {
+            cards = cards.filter((card) => properties.cardCondition(card, context));
+        }
+        events.push(event);
+        this.selectCard(event, additionalProperties, cards, new Set());
+
+        return events;
+    }
+
+    private getNumCards(numCards: Derivable<number>, context: AbilityContext): number {
+        return typeof numCards === 'function' ? numCards(context) : numCards;
+    }
+
+    private getAmount(amount: Derivable<number>, context: AbilityContext): number {
+        return typeof amount === 'function' ? amount(context) : amount;
+    }
+
+    private shouldShuffle(shuffle: Derivable<boolean>, context: AbilityContext): boolean {
+        return typeof shuffle === 'function' ? shuffle(context) : shuffle;
+    }
+
+    private getDeck(player: Player): Card[] {
+        return player.drawDeck;
+    }
+
+    private selectCard(event: DeckSearchEvent, additionalProperties: unknown, cards: Card[], selectedCards: Set<Card>): void {
```

Comment:
Is there a specific reason for `additionalProperties` to be `unknown` instead of `any`? Not sure when `unknown` is typically used